indico/modules/events/registration/models/registrations.py

```diff
@@ -488,6 +488,11 @@ def update_state(self, approved=None, paid=None, rejected=None, _skip_moderation
                 self.state = RegistrationState.pending
             elif paid is False and payment_required:
                 self.state = RegistrationState.unpaid
+        elif self.state == RegistrationState.rejected:
+            if rejected is False and payment_required is False and moderation_required:
```

Comment:
the `payment_required is False` check is not needed/correct here.

PS: can you add `bool()` around `self.price` above? while debugging i realized that `payment_required` is the price and not `True` ;x